### ISSUE TITLE
Terragrunt and TFLint above 0.50.0 breaking behavior

### ISSUE THREAD

**Issue by velkovb** [gruntwork-io/terragrunt#3669]
## Describe the enhancement

TFLint introduced a new way of handling values for undeclared variables in 0.50.0 - https://github.com/terraform-linters/tflint/pull/1941

This directly affects the Terragrunt hook integration as it does the following:

`tflint --config ./.terragrunt-cache/ellCj9USNaAqucT2-U5SmL6rf1A/Gu7P37gZ1L3NBihh7GJE9nHhW38/.tflint.hcl --chdir ./.terragrunt-cache/ellCj9USNaAqucT2-U5SmL6rf1A/Gu7P37gZ1L3NBihh7GJE9nHhW38 --var=test=foo` 

With Terragrunt release of `0.69.0` which bumped TFLint from 0.47.0 to 0.50.3 this started breaking our projects. We do generate a lot of metadata and automatic variables that are not used by every project. https://github.com/terraform-linters/tflint/discussions/2080#discussioncomment-11584584

The way that variables are passed to TFLint doesn't match the way they are passed to OpenTofu/ Terraform. For TFLint it seems to be with `--var` while for the latter it is `TF_VAR_*` which has no issues. If we pass OpenTofu/Terraform with `--var` it will error out as well.

Would be great if we can align both approaches and pass variables from input to TFLint as TF_VAR_* as well. If that cannot be done, passing a `.tfvars` file should work as well. I would guess the magic needs to happen somewhere here - https://github.com/gruntwork-io/terragrunt/blob/407f65fdb294559f3f70ac873aae5a41715d68dd/tflint/tflint.go#L154

Not sure if this is a bug or enhancement

## Additional context

Things you might want to address include:

- [ ] Changes required.
- [ ] Implications of the feature.
- [ ] Alternatives considered.
- [ ] Level of effort.

## PoC (Proof of Concept)

Link to a Proof of Concept if you have one:

- [ ] [PoC]()

Including a PoC can help others understand the feature better and implement it faster.

## RFC Not Needed

- [ ] I have evaluated the complexity of this enhancement, and I believe it does not require an RFC.

**Comment by yhakbar** [comment]
Hey @velkovb ,

Thanks for creating this issue. 

Unfortunately, given that there's a very straightforward workaround (calling tflint with the binary installed in PATH however users like), there isn't much incentive to adjusting how the built-in `tflint` hook works.

That being said, if this issue is important to you, would you consider contributing the code required to have `tflint` behave how you like? Even if you don't have a lot of Golang experience, we'd be happy to help you ramp up and collaborate with you on developing a solution.

Our focus is on delivering the features and updates that make Terragrunt ready for 1.0, so I'm jealously guarding the attention of the maintainers for that when I can. I hope you understand.

**Comment by yhakbar** [comment]
Ya, I was discussing the `--terragrunt-external-tflint` option with an older version of `tflint` installed so that you have the older behavior.

Please make sure to read [these docs](https://terragrunt.gruntwork.io/docs/community/contributing/) before you start contributing so that you have some guidance on how to get started.

The steps I generally recommend for new contributors is to start with the following:

1. Build Terragrunt locally, so that you can play with your own version of Terragrunt.
2. Run the relevant tests (you don't need to run the tests for the entire project, just the stuff that you're interested in changing).
3. Write a test that produces the outcomes you're looking for, and get those tests to fail the way you expect.
4. Write code that fixes the tests so that you have the new behavior you want.

I also heavily encourage you to join [the Discord](https://discord.gg/SPu4Degs5f) if you haven't. This can be a good place to ask questions about best practices, or to just become an active member of our community.

**Comment by velkovb** [comment]
@yhakbar I would guess that this logic is already implemented where the inputs are transformed for OpenTofu/ Terraform somewhere in the codebase. I could find it on my first attempt, could I ask for some directions where it is?

**Comment by wakeful** [comment]
hey,

Have you already read the main function where the `tflint` CLI is being constructed?

https://github.com/gruntwork-io/terragrunt/blob/main/tflint/tflint.go#L31

**Comment by velkovb** [comment]
We ended up with just calling TFLint with a custom hook, not the integrated one as we were not using the variables anyway. We have plans to enchance TFLint performance in the future and might revisit this than.

**Comment by NicolaSpreafico** [comment]
> We ended up with just calling TFLint with a custom hook, not the integrated one as we were not using the variables anyway. We have plans to enchance TFLint performance in the future and might revisit this than.

Hello,
I have the similar issue as you, precisely this:
```
We do generate a lot of metadata and automatic variables that are not used by every project
```

Right now I installed locally TFLINT 0.49.0 and I added the `"--terragrunt-external-tflint"` parameter in the TFLINT, but this is forcing me to remain on a fixed and older version of TFLint.

Can I ask you what is your solution about the custom hook?

Is there a way to tell TFLint to simply ignore this specific rule?
```
"A variable named "xxxx" was assigned, but the root module does not declare a variable of that name."
```
?

Thank you

**Comment by velkovb** [comment]
@NicolaSpreafico 

```hcl
terraform {
  before_hook "tflint" {
    commands = ["validate"]
    execute  = ["${get_repo_root()}/scripts/tflint.sh", "${get_repo_root()}/deployment/.tflint.hcl"]
  }
}
```

```bash
#!/bin/bash

tflint --init --config ${1}
```

**Comment by NicolaSpreafico** [comment]
@velkovb Thank you.

This is a mere alternative on using "_--terragrunt-external-tflint_" or does it actually solve the problem I mentioned of variable assigned and not deckared?

Thank you

**Comment by velkovb** [comment]
This doesn't use the Terragrunt integrated TFLint hook which actually passed the assigned variables so it doesn't have that problem. The issue here is it strips all variables even some that might have been actually used but that works for us for now. As described previously we plan on either trying to add an option to pass the variables as `TF_VAR_*` instead of with `--var` or doing something similar with the external hook.

**Comment by NicolaSpreafico** [comment]
Thank you, very clear.
In the .tflint.hcl did you set any particular configuration or it remains as before?

**Comment by velkovb** [comment]
@GergelyKalmar I understand your frustration as I was in the same situation. However, I do believe this is the way forward as tools evolve and we can't stay on an outdated version forever. Keeping in mind the licensing issue that would keep Terragrunt from upgrading builtin TFlint any further I truly believe that going external is the way to go.

You can check here what is the issue on TFLint side and why this is not exactly compatibility issue but rather a fixed behaviour. - https://github.com/terraform-linters/tflint/discussions/2080

@NicolaSpreafico Yes, updated my comment to reflect the correct script. Apologies for the confusion.

**Comment by NicolaSpreafico** [comment]
Hi @velkovb,
after my last reply I had the time to tweak a little your initial script, this is my current configuration:

root.hcl
```
before_hook "before_tflint" {
    commands = ["validate", "apply", "plan"]
    execute = ["${get_repo_root()}/scripts/tflint.sh", "${get_repo_root()}/project/live/.tflint.hcl", "${get_working_dir()}"]
  }
```

And this is the scripts/tflint.sh file
```
#!/bin/sh

# https://github.com/gruntwork-io/terragrunt/issues/3689
# https://github.com/gruntwork-io/terragrunt/issues/3669#issuecomment-2582486018

# When Terragrunt create the tflint command, if passes all the variable assignments as "var" parameter
# ---------------------------------------------------------------
# tflint --config ../../.tflint.hcl --chdir ./.terragrunt-cache/... --var=name1=value1 --var=name2=value2
# ---------------------------------------------------------------

# With Terragrunt release of 0.69.0, they bumped TFLint from 0.47.0 to 0.50.3
# With newest tflint versions, if there is no variable defined with that assignment, tflint exits with an error code:
# ---------------------------------------------------------------
# Failed to parse variables; <value for var.xxxx>:1,1-1: Value for undeclared variable; A variable named "xxxx" was assigned, but the root module does not declare a variable of that name., and 2 other diagnostic(s):
# 
# Error: Value for undeclared variable
# 
#   on <value for var.xxxx> line 1:
#   (source code not available)
# 
# A variable named "xxxx" was assigned, but the root module does not declare a variable of that name.
# ---------------------------------------------------------------

# Due to our Terragrunt structure, we define all the variables in different terragrunt.hcl commons files (global, same env for all modules or same module with different envs)
# Because of this, it is not guaranteed that all the assigned variables will be actually defined in all the terraform modules
# This is why we need to move to a way where we call tflint with a custom hook, and we do not pass the TF_VAR_ variables

# Check if tflint is not installed
if ! command -v tflint >/dev/null 2>&1
then
    echo "##########################################################################"
    echo "tflint is not installed"
    echo "Installation guide: https://github.com/terraform-linters/tflint?tab=readme-ov-file#installation"
    echo "##########################################################################"
    exit 1
else
    tflint --version
fi

TF_LINT_CONFIG_FILE="${1}"
TF_LINT_WORKING_DIR="${2}"

# echo "TF_LINT_CONFIG_FILE: ${TF_LINT_CONFIG_FILE}"
# echo "TF_LINT_WORKING_DIR: ${TF_LINT_WORKING_DIR}"

# Install plugins
tflint --config "${TF_LINT_CONFIG_FILE}" --init

# Process analysis
tflint --config "${TF_LINT_CONFIG_FILE}" --chdir "${TF_LINT_WORKING_DIR}" --recursive --color
```

**Comment by velkovb** [comment]
Nothing in the docs says that you can pass values for undefined variables. In practice, if you use Terragrunt in strict mode that should fail on `terragrunt validate-inputs` stage as well. If you are passing only values for defined variables you would have a problem. This is more of an edge case imho.

**Comment by gmspinheiro** [comment]
Agreed, especially since the Gruntwork's live example root.hcl file still suggests passing global inputs to all modules. It should not be expected all modules you might use have the same set of defined variables.

https://github.com/gruntwork-io/terragrunt-infrastructure-live-example/blob/d106b47ad1e22cae3a841cd495eae48e9bbe3a03/root.hcl#L70

**Comment by yhakbar** [comment]
You can still use an external version of tflint to bypass this issue, even if you don't want to avoid setting inputs that are unused.

At the end of the day, because of license issues in later versions of tflint, our options are to completely remove our integration with tflint, or to invest time in improving an integration with a tool that can't be upgraded further.

We've preserved the integration with the last fully open source version of the tool to support community members that depend on it, but we can't justify investing maintainer time in improving the integration, given the state of the project. If someone from the community wants to invest their time in improving the integration, we'll welcome a well tested PR.

**Comment by jmreicha** [comment]
There is some discussion happening in the opentofu project about how to address some of the linting issue https://github.com/opentofu/opentofu/issues/2213.

Unfortunately it doesn't look like this will be worked on until H2 2026 at the earliest so it seems calling the external version seems like the most viable solution for now.

**Comment by thisguycodes** [comment]
As of [v0.99.0](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.99.0) we are deprecating the internal tflint, and it will be fully removed in the future.

Though we still do automatic variable passing when calling an external tflint. Since this issue demonstrates this may not be desired behavior. This can be bypassed by specifying a full path to tflint rather than just `"tflint"` in your hook. I'll update the documentation to show this workaround.

I'm going to remove the `preserved` label. If this does not feel solved by provided means please feel free to comment and we can explore solutions, potentially re-preserving the issue.

**Comment by GergelyKalmar** [comment]
We have tried the workaround in the documentation, however, our tflint hooks keep failing because of the `TF_VAR_...` environment variables that terragrunt sets for every hook execution. Is there a way to control this and have terragrunt call out to tflint without setting these variables? We have a pretty standard setup with a single `inputs {}` block and it seems to be breaking the tflint calls no matter how we try.

**Comment by thisguycodes** [comment]
I'm happy to dig into this more. Is this the same as what you say at the top of the issue? Where if we're passing in variables that don't exist it errors?

You describe how the variables, or difference in the way it works between tflint and terraform, are causing problems. But can you be more specific?

**Comment by GergelyKalmar** [comment]
Well, it's definitely related to the issue as described above, but in any case, I'll try to first make then a simpler reproducer to be sure. Thank you for asking!

**Comment by GergelyKalmar** [comment]
I actually realized that it's only the error message that's confusing. If there is an error we get this:
```
11:01:28.886 INFO   Executing hook: tflint
1 issue(s) found:

Warning: Module should include an empty variables.tf file (terraform_standard_module_structure)

  on variables.tf line 1:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.13.0/docs/rules/terraform_standard_module_structure.md

11:01:28.949 ERROR  Error running hook tflint with message: Failed to execute "/usr/local/bin/tflint --color --config ./.tflint.hcl --chdir ." in .

exit status 2
11:01:28.949 ERROR  error occurred:

* Failed to execute "/usr/local/bin/tflint --color --config ./.tflint.hcl --chdir ." in .

  exit status 2
```

This suggests that there was some sort of an issue running tflint, while in fact it did run successfully, but found an issue.

**Comment by yhakbar** [comment]
Hey @GergelyKalmar ,

In general, a non-zero exit code in a hook is considered an error that needs addressing. Terragrunt doesn't try to do some inspection on the exit code of the underlying process it runs in the hook to determine if that process considers the non-zero exit code to indicate success or surfacing of additional information.

If you want to avoid that, you can use `bash -c "tflint || true"`.

**Comment by GergelyKalmar** [comment]
Yeah, that would stop CI/CD from marking the check failed though. The error message is more friendly when using the built-in tflint hook:
```
15:29:37.365 ERROR  [...] Error running hook tflint with message: Tflint found issues in the project. Check for the tflint logs.
```

This makes sense, however, the new error message is still confusing for this use case. Perhaps it could say something along the lines of "The hook encountered a non-zero exit status" instead of the current "Failed to execute".

**Comment by yhakbar** [comment]
That's a good idea! If you're open to creating a pull request to get a better error message there, that would be great. Explicitly telling the user what the exit code was seems useful too. Either way, please report it as a bad error message bug so it can be tracked.
